### PR TITLE
compactor: do not block compation when retention is taking too long

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
Currently, we perform compaction and apply retention in the same loop. Although we have a flag for not applying retention every time we perform compaction, we still see compaction getting blocked when processing some intensive delete requests(processed while applying retention).

This PR separates out the compaction and retention to run in a separate loop. I have added a table-locking feature to avoid compaction and retention from processing the same tables at a time. However, compaction and retention would treat locked tables differently, as explained below:
* When compaction sees a table is locked: It would skip the table and move on to the following table. However, before skipping, it would check if the table has any uncompacted files and increment the newly added counter called `loki_compactor_skipped_compacting_locked_tables_total` to track how often we are skipping compacting tables which have uncompacted files.
* When retention sees a table is locked: It would wait for the lock to be released since we can't skip any tables while processing delete requests.

**Special notes for your reviewer**:
* The check for tables with uncompacted files looks for count > 1 because initially, we did not support per tenant index in `boltdb-shipper`, so a table can have a single compacted multi-tenant index file. In a rare case where we have a single file which was supposed to be compacted away, it is okay to have a single uncompacted file for a while. The aim here is to not block compaction for too long in a large cell with too many uncompacted files.
* Retention only works on the compacted index, so we first compact down any uncompacted files while applying retention.

**Checklist**
- [x] Tests updated

### CODE REVIEW

**Comment by periklis** [comment]
@sandeepsukhani Does this only affect compaction when the Delete API is used?

**Comment by sandeepsukhani** [comment]
> @sandeepsukhani Does this only affect compaction when the Delete API is used?

Retention and compaction both get processed at the same time. Usually, delete requests which use line filters which are either touching too much data or slow to process are the ones blocking compaction.

**Comment by periklis** [comment]
> > @sandeepsukhani Does this only affect compaction when the Delete API is used?
> 
> Retention and compaction both get processed at the same time. Usually, delete requests which use line filters which are either touching too much data or slow to process are the ones blocking compaction.

Rewording my question a bit: The issue solved by this PR is not affecting compactors that simply apply retention stream delete rules when the Delete API is completely disabled, right?

**Comment by sandeepsukhani** [comment]
> > > @sandeepsukhani Does this only affect compaction when the Delete API is used?
> > 
> > 
> > Retention and compaction both get processed at the same time. Usually, delete requests which use line filters which are either touching too much data or slow to process are the ones blocking compaction.
> 
> Rewording my question a bit: The issue solved by this PR is not affecting compactors that simply apply retention stream delete rules when the Delete API is completely disabled, right?

Sorry, I made a mistake in my earlier response. Retention and delete request processing happen at the same time. The fix is for the issue, which happens only when there are intensive delete requests. If you have disabled retention or just the deletion, there should not be any issue blocking compaction.

**Review by ashwanthgoli — APPROVED** [review]
lgtm

File: pkg/compactor/metrics.go [hunk 1]
```diff
@@ -8,33 +8,36 @@ import (
...
 	statusSuccess = "success"
+
+	lblWithRetention = "with_retention"
 )
 
 type metrics struct {
 	compactTablesOperationTotal           *prometheus.CounterVec
-	compactTablesOperationDurationSeconds prometheus.Gauge
-	compactTablesOperationLastSuccess     prometheus.Gauge
+	compactTablesOperationDurationSeconds *prometheus.GaugeVec
```
[the remark this answers is not in the document]
Comment (sandeepsukhani) [reply pkg/compactor/metrics.go:17]: makes sense, I will do it in the same PR.

[the remark this answers is not in the document]
Comment (sandeepsukhani) [reply pkg/compactor/metrics.go:17]: I think I need to think more if we need it. I will do it in a separate PR.

**Pull request merged**